### PR TITLE
Throw exceptions on SQL errors

### PR DESCRIPTION
**Pull request by cedric-anne**
## Checklist before requesting a review

- [x] I have read the CONTRIBUTING document.
- [x] I have performed a self-review of my code.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] This change requires a documentation update.

## Description

Since we did not have a centralized error catching in the past, we were not able to throw exceptions when a SQL error occurred. They were just logged and were not stopping the code execution, and this could have the effect of triggering hard-to-debug errors in subsequently executed code.

Now we have a centralized error catching system, we can throw exceptions when a SQL error is triggered. It would result in displaying the error page. For instance, if I force a SQL querry error in the impact analysis tab:
image

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -122,7 +122,7 @@ The present file will list all changes made to the project; according to the
 - Usage of `DBmysql::query()`, `DBmysql::queryOrDie()` method are prohibited to ensure that legacy unsafe DB are no more executed.
   Building and executing raw queries using `DBmysql::request()`, `DBmysqlIterator::buildQuery()` and `DBmysqlIterator::execute()` methods is also prohibited.
   To execute DB queries, either `DBmysql::request()` can be used to craft query using the GLPI query builder,
-  either `DBmysql::doQuery()`/`DBmysql::doQueryOrDie()` can be used for safe queries to execute DB query using a self-crafted a SQL string.
+  either `DBmysql::doQuery()` can be used for safe queries to execute DB query using a self-crafted a SQL string.
```
Comment (trasher) [inline CHANGELOG.md:125]:
```suggestion
  `DBmysql::doQuery()` can be used for safe queries to execute DB query using a self-crafted a SQL string.
```

**Comment by cedric-anne** [comment]
E2E tests are failing due to the SQL error reported in #18175.

**Review by trasher — APPROVED** [review]

**Review by AdrienClairembault — COMMENTED** [review]
With these changes, should `Toolbox::logSqlError()` and the associated `sql-errors.log` log file be depreciated/removed ?

They are no longer used by Glpi's core.

**Comment by cedric-anne** [comment]
> With these changes, should `Toolbox::logSqlError()` and the associated `sql-errors.log` file be depreciated/removed ?
>
> With the new exception, SQL errors end up directly in the `php-errors-log` file. I don't see a use case for manually calling `Toolbox::logSqlError()`.

`sql-errors.log` still contains SQL warnings and SQL debug info. We could change it, but it is out of the scope of the current PR.

**Comment by trasher** [comment]
Any reason this one is still draft, or should me merge (e2e error is not a blocker)?

**Comment by cedric-anne** [comment]
> Any reason this one is still draft, or should me merge (e2e error is not a blocker)?

The e2e tests fails due to an SQL error that is now blocking, see #18175. We have to fix this SQL error before being able to merge the current PR.

**Comment by cedric-anne** [comment]
I added a commit to quickly adapt the install/update process.

File: install/install.php [hunk 2]
```diff
@@ -288,102 +285,70 @@ public function __construct($dbh)
...
+        $databasename,
+        use_timezones: $timezones_requirement->isValidated(),
+        log_deprecation_warnings: false,
+        use_utf8mb4: true,
+        allow_datetime: false,
+        allow_signed_keys: false
+    );
+
+    if ($success) {
+        echo "<p>" . __('Initializing database tables and default data...') . "</p>";
```
Comment (trasher) [inline install/install.php:331]: We probably need string escaping here (for all `echo`)
```suggestion
        echo "<p>" . __s('Initializing database tables and default data...') . "</p>";
```

File: src/Update.php [hunk 3]
```diff
@@ -259,7 +259,16 @@ function (string $flag) {
...
-            $migration_specs['function']();
+
+            try {
+                $migration_specs['function']();
+            } catch (\Throwable $e) {
+                $this->migration->displayError(sprintf(
+                    __('An error occurred during the update. The error was: %s'),
+                    $e->getMessage()
+                ));
+                die(1);
```
Comment (trasher) [inline src/Update.php:270]: Aren't we supposed to drop all `die` and similar calls? (I do not have in mind the process)

File: install/install.php (hunk 2 above)
Comment (cedric-anne) [reply install/install.php:331]: I planned to do it in another PR, with all other install/update scripts echo.

File: src/Update.php (hunk 3 above)
Comment (cedric-anne) [reply src/Update.php:270]: Yes, but this one requires many changes, and this PR is already big enough.

Comment (trasher) [reply src/Update.php:270]: ok

File: install/install.php (hunk 2 above)
Comment (trasher) [reply install/install.php:331]: ok

**Review by trasher — APPROVED** [review]

**Pull request merged**